### PR TITLE
Indexer

### PR DESCRIPTION
**Pull request by ethanleifer**

### CODE REVIEW

**Comment by zackgow** [comment]
1. Right off the bat got an error when following the README:
```
Current Python version (3.11.0) is not allowed by the project (3.10).
Please change python executable via the "env use" command.
```
2. env files we should probably store in cloud storage. Then we can just have a `MAKEFILE` to fetch new envs

File: packages/indexer/indexer/chain_mapper.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+@dataclass
+class CosmosAPI:
+    url: str
+    archive: bool = False
+
+
+chain_mapping: List[CosmosChain] = [
+    CosmosChain(
+        chain_id="secret-4",
+        min_block_height=7284419,
```
Comment (zackgow) [inline packages/indexer/indexer/chain_mapper.py:22]: I don't like that these are hardcoded. What's the plan for when we move to `secret-5`?

File: packages/indexer/indexer/chain_mapper.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        apis=[
+            CosmosAPI(url="https://secret-4.api.trivium.network:1317"),
```
Comment (zackgow) [inline packages/indexer/indexer/chain_mapper.py:24]: Same thing here - no hardcoding of endpoints

File: packages/indexer/indexer/chain_mapper.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        ],
+    ),
+    CosmosChain(
+        chain_id="jackal-1",
```
Comment (zackgow) [inline packages/indexer/indexer/chain_mapper.py:28]: We talked about this in Discord, but I think it is a bad design decision to jam all different chains into a single indexer.

File: packages/indexer/indexer/chain_mapper.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        min_block_height=1,
+        apis=[
+            CosmosAPI(url="https://api.jackalprotocol.com"),
+            CosmosAPI(url="https://jackal-api.polkachu.com"),
+            CosmosAPI(url=" https://api.jackal.nodestake.top"),
+        ],
+    ),
+    CosmosChain(
```
Comment (zackgow) [inline packages/indexer/indexer/chain_mapper.py:36]: Same thing

File: packages/indexer/indexer/data.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    session: aiohttp.ClientSession,
+    api: CosmosAPI,
+    height: str = "latest",
+    max_retries=5,
+) -> dict | None:
+    retries = 0
+    while retries < max_retries:
+        try:
+            async with session.get(
+                f"{api.url}/cosmos/base/tendermint/v1beta1/blocks/{height}"
```
Comment (zackgow) [inline packages/indexer/indexer/data.py:15]: To keep things clean, maybe at the top of the script we have something like: `block_endpoint = f`{api.url}/cosmos/base/tendermint/v1beta1/blocks/`. In our current codebase, there has been several times where I've forgotten to update endpoints when they were mixed throughout the codebase. Keeping it at the top would be organized. 

Or if you are using objects, then we might as well just have that defined on the `CosmosAPI` object.

File: packages/indexer/indexer/chain_mapper.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+@dataclass
+class CosmosChain:
+    chain_id: str
+    min_block_height: int
+    apis: List["CosmosAPI"] = field(default_factory=list)
+    current_api_index: int = 0
+
+
+@dataclass
+class CosmosAPI:
```
Comment (zackgow) [inline packages/indexer/indexer/chain_mapper.py:14]: I think we need to rename these

File: packages/indexer/indexer/db.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    await pool.execute(
+        """
+        DROP TABLE IF EXISTS raw CASCADE;
+        DROP TABLE IF EXISTS blocks CASCADE;
+        DROP TABLE IF EXISTS txs CASCADE;
+        """
+    )
+
+
+async def create_tables(pool: asyncpg.pool):
```
Comment (zackgow) [inline packages/indexer/indexer/db.py:19]: These queries I think should exist in `.sql` files and are then read-in. In-line makes it harder to read what the code is doing.

File: packages/indexer/indexer/db.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                VALUES (
+                    NEW.height,
+                    NEW.chain_id,
+                    (NEW.block->'block'->'header'->'time')::TEXT::timestamp without time zone,
+                    (NEW.block->'block_id'->>'hash')::TEXT,
+                    (NEW.block->'block'->'header'->>'proposer_address')::TEXT
+                );
+                
+                        
+                FOR tx_responses IN SELECT * FROM jsonb_array_elements(NEW.txs->'tx_responses')
```
Comment (zackgow) [inline packages/indexer/indexer/db.py:93]: This syntax is strange to me

File: packages/indexer/indexer/db.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            $$ LANGUAGE plpgsql;
+            CREATE OR REPLACE TRIGGER raw_insert
+            BEFORE INSERT
+            ON raw
+            FOR EACH ROW EXECUTE PROCEDURE parse_raw();
+            """
+        )
+
+
+async def upsert_block(pool: asyncpg.pool, block: dict, txs: dict):
```
Comment (zackgow) [inline packages/indexer/indexer/db.py:127]: This function is super clean

File: packages/indexer/indexer/db.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        current_block = int(block_data["block"]["header"]["height"])
+        async with pool.acquire() as conn:
+            rows = await conn.fetch(
+                """
+                SELECT height FROM blocks WHERE chain_id = $1
+                """,
+                chain.chain_id,
+            )
+            heights = [row[0] for row in rows]
+            for height in range(chain.min_block_height, current_block):
```
Comment (zackgow) [inline packages/indexer/indexer/db.py:166]: My concern is what if there is missing data in-between for some reason?

For example, say we have blocks `[1,2,3,4,5]`
then later end up in a state where we have `[1, 2, 3, 4, 5, 10, 11, 12, 13]` but the current blockheight is `20`. We want to make sure we don't query any blocks that we already have.

File: packages/indexer/indexer/main.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                        height=current_block,
+                        pool=pool,
+                        session=session,
+                        block_data=block_data,
+                    ):
+                        print(
+                            f"processed live block {current_block} - {chain.chain_id}"
+                        )
+                    else:
+                        # should we save this to db?
```
Comment (zackgow) [inline packages/indexer/indexer/main.py:68]: We could have an error table but I'm afraid of things continually just piling up in there. With GCP we can enable logging and just search those logs. There are some tools to make this easier but let's ignore for now.

File: packages/indexer/indexer/test.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import asyncio
+import uuid
+
+
```
Comment (zackgow) [inline packages/indexer/indexer/test.py:4]: We need real tests but fine for now.

**Comment by zackgow** [comment]
All of the triggers to parse things out does make me wonder if we should just ingest the raw files, stream them to GCP cloud storage, and use Cloud DataFlow via Apache Beam or something like Bytewax to parse that stream in real-time, and then ingest the real-time stream into a real OLAP database like BigQuery or DuckDB.

File: packages/indexer/indexer/chain_mapper.py (hunk 1 above)
Comment (ethanleifer) [reply packages/indexer/indexer/chain_mapper.py:22]: well we will have to change api endpoints for secret-5 anyways..its a different chain

File: packages/indexer/indexer/chain_mapper.py (hunk 2 above)
Comment (ethanleifer) [reply packages/indexer/indexer/chain_mapper.py:24]: agreed should all be in a `.env`

File: packages/indexer/indexer/data.py (hunk 5 above)
Comment (ethanleifer) [reply packages/indexer/indexer/data.py:15]: I like this thought

File: packages/indexer/indexer/chain_mapper.py (hunk 6 above)
Comment (ethanleifer) [reply packages/indexer/indexer/chain_mapper.py:14]: to what?

File: packages/indexer/indexer/db.py (hunk 8 above)
Comment (ethanleifer) [reply packages/indexer/indexer/db.py:93]: It was bizarre to me too but think of it like: for each row we need to normalize the data based on the json. 

I can store the select in a variable tho to make it look cleaner

File: packages/indexer/indexer/db.py (hunk 9 above)
Comment (ethanleifer) [reply packages/indexer/indexer/db.py:127]: im proud of this one lol

File: packages/indexer/indexer/db.py (hunk 10 above)
Comment (ethanleifer) [reply packages/indexer/indexer/db.py:166]: it accounts for this:

min height in your example is 1 so you iterate from 1-20 removing those blocks in that list (look at sudocode below)

```
for every possible height in blockchain (from min height to current height):
    if this height hasn't queried
        upsert
```

File: packages/indexer/indexer/main.py (hunk 11 above)
Comment (ethanleifer) [reply packages/indexer/indexer/main.py:68]: it seems dumb to keep  reproccess blocks that keep causing errors imo

**Comment by ethanleifer** [comment]
We can brainstorm on that on Monday. I agree that something like that could work well but this is def simpler from an architectural standpoint

File: packages/indexer/indexer/chain_mapper.py (hunk 1 above)
Comment (zackgow) [reply packages/indexer/indexer/chain_mapper.py:22]: Right, hence they should be env variables

File: packages/indexer/indexer/chain_mapper.py (hunk 6 above)
Comment (zackgow) [reply packages/indexer/indexer/chain_mapper.py:14]: Not sure, but `CosmosAPI` isn't accurate.

File: packages/indexer/indexer/db.py (hunk 10 above)
Comment (zackgow) [reply packages/indexer/indexer/db.py:166]: I still think we need to take some set difference or something. What if there are tens of millions of blocks? Seems silly to iterate over the entire list.

File: packages/indexer/indexer/main.py (hunk 11 above)
Comment (zackgow) [reply packages/indexer/indexer/main.py:68]: Yeah agreed.

File: packages/indexer/indexer/data.py (hunk 5 above)
Comment (zackgow) [reply packages/indexer/indexer/data.py:15]: Another reason, is not all chains will be on `v1beta`. This depends on what version of tendermint they are running.

File: packages/indexer/indexer/db.py (hunk 10 above)
Comment (ethanleifer) [reply packages/indexer/indexer/db.py:166]: Its a generator so it yields blocks as they are needed, this is more efficient than doing all of the computation at the beginning

**Review by zackgow — COMMENTED** [review]
Add a .gitignore file so you don't accidentally commit .csv, .txt, etc.

**Comment by ethanleifer** [comment]
Still to think about list: 
- when we fail to process a block / log due to some data issue, how r we logging/storing it?
- on conflict stuff 
- update case for triggers (what about delete?)
- what are we doing if db pool or aiohttp session break? (currently this stops the script) 
- more performance improvements in parsing:
      - upsert speeds using db transactions instead of tons of separate executes
      - asyncio.gather on multiple await calls
- chain-registery json file for multiple chains, (this has chain_id and other information in it we might want to use) ex: min block height, some address prefix stuff that might be useful etc.. 
- also do we care about block data now? should we make this an env variable? its fully decoupled now so not a big deal if you dont want it. if we want to enable it for Wed I need to make a few code changes

File: .gitignore [hunk 13]
```diff
@@ -1,2 +1,3 @@
 .env
```
Comment (zackgow) [inline .gitignore:1]: add .csv, .txt., etc. to this

File: packages/indexer/indexer/db.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+                    select height, COALESCE(height - LAG(height) over (order by height), -1) as difference_per_block, chain_id
+                    from raw_txs
+                    where chain_id = $1
+                ) as dif
+                where difference_per_block <> 1
+        """,
+            chain.chain_id,
+        )
+
+        # the above query returns the difference between the two blocks that exists
```
Comment (zackgow) [inline packages/indexer/indexer/db.py:176]: All of these comments should be in the doc string

**Comment by zackgow** [comment]
I'm merging this for now. Moving forward let's push for smaller PRs. We need to add some tests before anything else.

**Pull request merged**